util/rowcodec/common.go

```diff
@@ -240,3 +245,109 @@ func IsNewFormat(rowData []byte) bool {
 func FieldTypeFromModelColumn(col *model.ColumnInfo) *types.FieldType {
 	return col.FieldType.Clone()
 }
+
+// ColData combines the column info as well as its datum. It's used to calculate checksum.
+type ColData struct {
+	*model.ColumnInfo
+	Datum *data.Datum
+}
+
+// Encode encodes the column datum into bytes for checksum. If buf provided, append encoded data to it.
+func (c ColData) Encode(buf []byte) ([]byte, error) {
+	return appendDatumForChecksum(buf, c.Datum, c.GetType())
+}
+
+// RowData is a list of ColData for row checksum calculation.
+type RowData struct {
+	Cols []ColData
+	Data []byte
+}
+
+// Len implements sort.Interface for RowData.
+func (r RowData) Len() int { return len(r.Cols) }
+
+// Less implements sort.Interface for RowData.
+func (r RowData) Less(i int, j int) bool { return r.Cols[i].ID < r.Cols[j].ID }
+
+// Swap implements sort.Interface for RowData.
+func (r RowData) Swap(i int, j int) { r.Cols[i], r.Cols[j] = r.Cols[j], r.Cols[i] }
+
+// Encode encodes all columns into bytes (for test purpose).
+func (r *RowData) Encode() ([]byte, error) {
+	var err error
+	if len(r.Data) > 0 {
+		r.Data = r.Data[:0]
+	}
+	for _, col := range r.Cols {
+		r.Data, err = col.Encode(r.Data)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return r.Data, nil
+}
+
+// Checksum calculates the checksum of columns. Callers should make sure columns are sorted by id.
+func (r *RowData) Checksum() (checksum uint32, err error) {
+	for _, col := range r.Cols {
+		if len(r.Data) > 0 {
+			r.Data = r.Data[:0]
+		}
+		r.Data, err = col.Encode(r.Data)
+		if err != nil {
+			return 0, err
+		}
+		checksum = crc32.Update(checksum, crc32.IEEETable, r.Data)
+	}
+	return checksum, nil
+}
+
+func appendDatumForChecksum(buf []byte, dat *data.Datum, typ byte) (out []byte, err error) {
```

Comment:
It looks for all columns, concatenate the bytes together, does not consider the bytes length.